src/printer.js

```diff
@@ -2977,6 +2971,50 @@ function hasLeadingOwnLineComment(text, node) {
   return res;
 }
 
+// A "leadingComment" attached to the top of the first token in the
+// return's argument, which would originally be in parens,
+// could trigger an ASI error where the parens are removed
+// and the comment bumped to the same line as the `return`,
+// while the contents of the return value are on the next line (oops!).
+//
+// This recurses the return argument, looking for the first token
+// (the leftmost leaf node) and, if it (or its parents) has any
+// leadingComments, wraps the return in parens.
+function returnArgumentHasLeadingComment(argument) {
+  const hasLeadingComment = (comments) => (
+    comments && comments.some(comment => comment.leading)
+  );
+  const getLeftSide = (node) => {
```

Comment:
(copying comments from previous thread)

feel a little weird defining this function in here, but it's very useful to have and wouldn't necessarily apply more broadly.
 
I also considered a switch/case explicitly listing the node types corresponding to each relevant node attribute; figured to go with this since it's more concise, but could have subtle bugs since only the top node's type is verified, and ~any expression could be inside its stack...

for example, if someone threw an AssignmentExpression in there as the "object" of a MemberExpression ((a=b).b), it would traverse the AssignmentExpression too, since it has a left attribute.

I don't think that's a problem, but figured it should be mentioned